New extended resources kublet flag

sethp-nr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Enables kubeadm configured nodes to start up advertising a specific set
of extended resources, instead of needing to wait for the node to
register in the API and send a specific PATCH to the node's status.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
None that I could find, but both myself and one other person have raised the question in the sig-node slack channel.

**Special notes for your reviewer**:

I ignored the very good advice to not implement new kubelet flags because it's not clear to me how to define which subset of nodes get which KubeletConfiguration(s).

That said, I'm happy to learn more about component configs if you would help guide me in moving this config there and attaching it to the right nodes via `kubeadm`.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Adds a --node-extended-resources flag to instruct the kubelet to register the node with statically defined extended resoures
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mtaufen commented:
/assign @dashpole

dashpole commented:
Can you say a bit more about your use-case?  Why doesn't sending a PATCH work?

sethp-nr commented:
Sure! We're using cluster-api to manage our clusters, which means that there's not necessarily a human operator standing by to issue the PATCH when a new node joins the cluster. And in the cases where there is a human being, we often fail to issue the right command 😄

What cluster-api does give us is a way to plumb extra kubelet args through the `kubeadm` config applied to those machines to ensure they come into the cluster with the right set of resources.

The alternative we considered was a NoSchedule taint + controller that watches for some particular node label, and issues the PATCH and removes the NoSchedule taint. But since the `kubelet` already has flags for taints and the node label, it made sense to us (and would be operationally simpler) for it to also understood its own extended resources.

dashpole commented:
If a pod requests an extended resource, but the node object doesn't have it, will it still be scheduled?  I'm trying to figure out why the taint is required

sethp-nr committed:
Reword comment and error message

Incorporating review feedback. I'm leaving the "non-negative" in there
in lieu of "positive" because I want to allow for the possibility of
bootstrapping a node with a zero value for an extended resource.

neolit123 commented on cmd/kubelet/app/options/options.go:
```diff
@@ -243,6 +247,13 @@ func ValidateKubeletFlags(f *KubeletFlags) error {
 	}
 
+	for k, v := range f.NodeExtendedResources {
+		// This is possibly a little more restrictive then necessary, but all pure integers are valid Quantities
```
then -> than

neolit123 commented on cmd/kubelet/app/options/options.go:
```diff
@@ -243,6 +247,13 @@ func ValidateKubeletFlags(f *KubeletFlags) error {
+	for k, v := range f.NodeExtendedResources {
+		// This is possibly a little more restrictive then necessary, but all pure integers are valid Quantities
+		if i, err := strconv.ParseInt(v, 10, 64); err != nil || i < 0 {
+			return fmt.Errorf("node extended resources must be whole 64-bit signed (but non-negative) integers. Instead we saw %s=%s", k, v)
```
excluding the personallization "we".
"whole" numbers implies non-negative integers, possibly adjust.

proposing:

> node extended resources must be 64bit signed (but positive) integers; saw %s=%s

neolit123 commented:
@sethp-nr thanks for the PR, was there a discussion about this in the cluster-api or kubernetes/kubernetes issue tracker already?

> What cluster-api does give us is a way to plumb extra kubelet args through the kubeadm config applied to those machines to ensure they come into the cluster with the right set of resources.

alternatively, if are willing to use https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm for v1alpha2 of cluster-api.

i think it will allow you to apply a PATCH as a post `kubeadm init/join` step using cloud-init scripts.

mattjmcnaughton commented:
@sethp-nr welcome to the k8s community :)

Could I trouble you to link the slack conversation? This seems like a non-trivial feature to add, so I just want to make sure its been given the proper discussion before we get too far into the code.

cc @derekwaynecarr @dchen1107 would we want a KEP for this feature? I'm not sure so wanted to escalate to y'all :)

sethp-nr committed:
New extended resources kublet flag

Enables kubeadm configured nodes to start up advertising a specific set
of extended resources, instead of needing to wait for the node to
register in the API and send a specific PATCH to the node's status.
